File: moderate/longestlines.rb
lol = []
n = 0
File.open(ARGV[0]).each_line do |line|
  if n == 0
    n = line.to_i
    next
  end
  line.chomp!
  (0...n).each do |i|
    if i == lol.length
      lol[i] = line
      break
    end
    lol[i], line = line, lol[i] if line.length > lol[i].length
  end
end
puts lol.join("\n")

File: moderate/longestlines.pl
use strict;
my $n = 0;
my @l;
open( INFILE, "<", $ARGV[0] )
  or die("Cannot open file $ARGV[0] for reading: $!");
while ( my $line = <INFILE> ) {
    chomp($line);
    if ( $n == 0 ) {
        $n = $line;
        next;
    }
    for ( my $i = 0 ; $i < $n ; $i++ ) {
        if ( $i == scalar @l ) {
            push @l, $line;
            last;
        }
        ( @l[$i], $line ) = ( $line, @l[$i] )
          if ( length $line > length @l[$i] );
    }
}
printf( "%s\n", join( "\n", @l ) );
close(INFILE);
